support multiple reference genomes 2nd try

khzhu opened the pull request:
# What? Why?
Fix 
#5652 
Changes proposed in this pull request:
- [x] remove **cytoband** & **length** from the **_gene_** table
- [x] replace **_entrez_gene_id_** with **_genetic_entity_id_** in **_reference_genome_gene_** table to support gene isoforms
- [x] add REFERENCE_GENOME_ID to the **_cancer_study_** table
  - [x] if no REFERENCE_GENOME_ID specified in the study meta file use default GRCh37/hg19
- [x] add a new API endpoint **_reference-genome-genes_** for querying gene related annotation such as cytoband/exonic length/chromosome from **_reference_genome_gene_** table
- [x] for mutations, use the **_CHR_** in the mutation_event table instead of calculating chromosome from the **_cytoband_** recorded in the gene table
- [x] add a database constraint to ensure one species per portal installation

# Checks
- [ ] Runs on Heroku.
- [ ] Follows [7 rules of great commit messages](http://chris.beams.io/posts/git-commit/). For most PRs a single commit should suffice, in some cases multiple topical commits can be useful. During review it is ok to see tiny commits (e.g. Fix reviewer comments), but right before the code gets merged to master or rc branch, any such commits should be squashed since they are useless to the other developers. Definitely avoid [merge commits, use rebase instead.](http://nathanleclaire.com/blog/2014/09/14/dont-be-scared-of-git-rebase/)
- [ ] Follows the [Google Style Guide](https://github.com/google/styleguide).
- [ ] If this is a feature, the PR is to rc. If this is a bug fix, the PR is to master.

# Any screenshots or GIFs?
If this is a new visual feature please add a before/after screenshot or gif
here with e.g. [GifGrabber](http://www.gifgrabber.com/).
#

sheridancbio commented:
Hi Kelsey. I'm going to start reviewing this. It looks good, but it is pretty big. Could you give a few suggestions on what would be the best classes / configuration to start looking at to understand the important changes? (Advice on how to navigate through the PR and understand the important stuff early in the process) Thanks.

sheridancbio commented:
I see several files with name "pom.xml.new-version" checked in. Aren't these artifacts of the maven build script? I think these can be dropped from the PR.

sheridancbio commented:
dcx

sheridancbio commented:
I see this image file added, but no references to it. I wonder... is there another PR for frontend changes connected to this work?
portal/src/main/webapp/images/uhn-logo.png

khzhu commented:
@sheridancbio , oops, should not be included, will remove it. Thanks!

khzhu commented:
@sheridancbio , please take a look at those red notes/circles which outlines those major changes made. 
![support-multiple-genomes-2nd-5652](https://user-images.githubusercontent.com/17375786/55174329-9eced000-5153-11e9-8b8b-e4c88bed1085.png)

khzhu commented:
@sheridancbio , yes, they should be removed as well. I used "git add -A" due to a large number of files changed, that picked some unwanted files as well. Sorry!

khzhu commented:
@sheridancbio , extra files removed and push the change back.

sheridancbio commented:
currently there are three service module unit tests failing in travis-ci:

khzhu commented:
@sheridancbio , will take a look. The PR passed all unit testing when I was testing. thanks!

khzhu commented:
@sheridancbio , pushed the change back, should fix those failed unit testing. Thanks!

khzhu commented:
hi @sheridancbio , please find updated schema diagram. I will update the code once I made changes. Thanks!
![support-multiple-genomes-5652-revised](https://user-images.githubusercontent.com/17375786/55636883-d829c000-5791-11e9-864c-c9131c6f798d.png)

khzhu commented:
Hi @sheridancbio , pushed the change back and is ready for the review. Thanks!

sheridancbio commented on core/src/main/java/org/mskcc/cbio/portal/dao/DaoReferenceGenomeGene.java:
```diff
@@ -0,0 +1,255 @@
+            else {
+                rows += updateGene(existingGene);
+            }
+            return rows;
```
again, I wonder whether we really need to tell the caller whether a change happened (versus no change was necessary). Making this function void would avoid the bother, or making it type boolean and having statements like `return pstmt.executeUpdate() == 1` or `return updateGene(existingGene)` seems more clear to me than returning an integer that then needs to be interpreted by the caller.

khzhu replied:
agree, changed to void.

sheridancbio commented on core/src/main/java/org/mskcc/cbio/portal/dao/DaoReferenceGenomeGene.java:
```diff
@@ -0,0 +1,255 @@
+     * @return Canonical Gene Object.
+     * @throws DaoException Database Error.
+     */
+    public ReferenceGenomeGene getGene(String hugoGeneSymbol, int referenceGenomeId) throws DaoException {
```
On my first readthrough of this code, this join logic means that this function will be blind to any records in reference_genome_gene that do not have a matching (by entrez_gene_id) record in the table gene. So I am thinking that it always will be the case that when records are added to reference_genome_gene, there will be a matching record in table gene to link to. (maybe I am just stating the obvious here)

I guess I might recommend that in the code which adds records to reference_genome_gene, we might want to add a check that a matching record exists in `gene`. (or maybe we already have a foreign key constraint between tables ... I have not gotten there yet)

khzhu replied:
yes, we have a foreign key constraint in the reference genome table links to the gene table, you will not be able to add to the reference gene table if no record found in the gene table, the beauty of foreign key :) .

sheridancbio commented on core/src/main/java/org/mskcc/cbio/portal/dao/DaoReferenceGenomeGene.java:
```diff
@@ -0,0 +1,255 @@
+            pstmt.setInt(2, referenceGenomeId);
+            rs = pstmt.executeQuery();
+            if (rs.next()) {
+                return extractGene(rs);
```
we use the same extractGene() function here and also in the previous function call which does not join tables. I guess I wonder: are we getting any additional information from the gene table when extracting this gene? If not, then why are we joining the tables?

Ah .. I think I see --- this function takes hugo gene symbol, and we are using that as a search criteria against the gene table. Maybe a more natural way to do this would be to first search the gene table with hugo gene symbol and resolve it to an entrez gene id. After that, you can call the prior function with the entrez_gene_id and the reference_genome_id.

Another issue here is that we use the gene_alias table sometimes to resolve hugo gene symbols. The way this is written, getGene(String hugoGeneSymbol, int referenceGenomeId) is only going to find exact matches to Hugo gene symbols in the gene table. Do we not need to resolve based on aliases sometimes?

khzhu replied:
no so sure if we any use case like querying based on gene aliases. 
ah, good idea. We could make two calls: call to the gene table for retrieving hugoGeneSymbol followed by the one to the reference_genome_gene table. It might be faster this way since both tables are cached.

sheridancbio commented on core/src/main/java/org/mskcc/cbio/portal/dao/DaoReferenceGenomeGene.java:
```diff
@@ -0,0 +1,255 @@
+        ResultSet rs = null;
+        try {
+            con = JdbcUtil.getDbConnection(DaoReferenceGenomeGene.class);
+            pstmt = con.prepareStatement("DELETE FROM `reference_genome_gene` WHERE ENTREZ_GENE_ID=? AND REFERENCE_GENOME_ID=?");
```
we probably need functionality to be added to DaoGene ... which will delete all linked ReferenceGenomeGene records for an entrez_gene_id when you delete that entrez_gene_id from the gene table.

khzhu replied:
we do not need one here since I had a **_cascade_** deletion foreign key i.e. when a gene deleted all reference genome genes will be automatically deleted as well.

sheridancbio commented on core/src/main/java/org/mskcc/cbio/portal/dao/DaoReferenceGenomeGene.java:
```diff
@@ -0,0 +1,255 @@
+     * Deletes all Reference Genome Gene Records in the Database.
+     * @throws DaoException Database Error.
+     *
+     * @deprecated only used by deprecated code, so deprecating this as well.
```
Let's not give birth to deprecated code ... consider deleting this if it is not needed.
Having said that ... if we have a deleteAllRecords() function in DaoGene, then that function would need to call this function I think.

khzhu replied:
agree. Will remove it.

sheridancbio commented on core/src/main/java/org/mskcc/cbio/portal/dao/JdbcUtil.java:
```diff
@@ -56,7 +56,7 @@
      */
     public static DataSource getDataSource() {
         if (dataSource == null) {
-            dataSource = new JdbcDataSource();
```
We need to be careful here. There is code which wires beans appropriately during spring startup for determination of the data source based on portal.properties and application-context files. We use a shared data source within tomcat when running multiple instances of cBioPortal within a single tomcat application. It seems like this code change may hardwire a connection pool through the apache BasicDataSource library onto each running tomcat web application.
We will do some testing of this. Maybe @n1zea144 can comment on this code change.

khzhu replied:
good points, if you need to run multiple portal instances in a single tomcat container. Will revert the change.

khzhu replied:
@sheridancbio , only with one change made to JdbcDataSource constructor, to replace hard code mysql driver class name with value read from portal.properties file as the following:
from
> this.setDriverClassName("com.mysql.jdbc.Driver");

to

> String mysqlDriverClassName = dbProperties.getDbDriverClassName();
> this.setDriverClassName(mysqlDriverClassName);

sheridancbio replied:
reminder to self : let's test this code in tomcat

sheridancbio commented on core/src/main/java/org/mskcc/cbio/portal/dao/JdbcUtil.java:
```diff
@@ -69,6 +69,28 @@ public static void setDataSource(DataSource value) {
+        String userName = dbProperties.getDbUser();
+        String password = dbProperties.getDbPassword();
+        String database = dbProperties.getDbName();
+        String url ="jdbc:mysql://" + host + "/" + database +
```
The overriding of userName, password, and other connection string terms is also being handled more subtly than may first appear here. I think this change also needs careful testing to make sure that properties passed in either through the portal.properties files or through java "-Dproperty_name=property_value" clauses will still work correctly.

khzhu replied:
will revert the change.

khzhu commented:
@sheridancbio , thank you so much for the feedback. I will go through each of your comments/suggestions over the weekend.

ao508 commented on business/src/main/resources/org/mskcc/cbio/portal/persistence/GenePanelMapperLegacy.xml:
```diff
@@ -70,8 +70,6 @@
         <result property="entrezGeneId" column="ENTREZ_GENE_ID"/>
         <result property="hugoGeneSymbol" column="HUGO_GENE_SYMBOL"/>
         <result property="type" column="TYPE"/>
-        <result property="cytoband" column="CYTOBAND"/>
```
@sheridancbio it looks like `reference_genome_gene` now stores the cytoband and exonic length

ao508 replied:
Note for ourselves: we will need to make the proper changes to our weekly gene updater tool model, db operations, etc. @n1zea144 @sheridancbio

khzhu replied:
cytoband differs from each genome build. In order to support multiple genome builds in a single portal, we need to have those gene information such as cytoband, exonic length, start, and end of gene being recorded in the reference genome table. And, cBioPortal is built around gene that is why a large number of files affected by this change.

ao508 commented on db-scripts/src/main/resources/adjust_col_size_to_utf8.sql:
```diff
@@ -0,0 +1,16 @@
+DROP PROCEDURE IF EXISTS adjust_col_size_to_utf8;
```
@khzhu So a while ago I thought to introduce a stored procedure to cleanup orphan records in certain tables but @ersinciftci made a valid point that doing so may lead to problems trying to debug issues - something along those lines. 

@ersinciftci can you comment on this?

ao508 replied:
@khzhu I also recently introduced a PR that changed all allele columns in `mutation` and `mutation_event` to `TEXT` datatypes and we are instead doing a partial index on `mutation_event.TUMOR_SEQ_ALLELE`.

khzhu replied:
with your PR, this no longer needed, will remove.

ao508 commented on db-scripts/src/main/resources/cgds.sql:
```diff
@@ -111,6 +111,20 @@ CREATE TABLE `type_of_cancer` (
   PRIMARY KEY (`TYPE_OF_CANCER_ID`)
 );
 
```
I see 2 references to dropping table `reference_genome_gene`

- [L85](https://github.com/cBioPortal/cbioportal/pull/5891/files?file-filters%5B%5D=.sql#diff-0146a84b4bc5ff8d2ac00911a327d25fR85)
- [L100](https://github.com/cBioPortal/cbioportal/pull/5891/files?file-filters%5B%5D=.sql#diff-0146a84b4bc5ff8d2ac00911a327d25fR100)

khzhu replied:
good catch, removed line 100.

ao508 commented on db-scripts/src/main/resources/migration.sql:
```diff
@@ -417,9 +418,11 @@ CREATE TABLE `reference_genome_gene` (
 );
 
-INSERT INTO reference_genome_gene (ENTREZ_GENE_ID, CYTOBAND, EXONIC_LENGTH, CHR, REFERENCE_GENOME_ID)
+INSERT INTO reference_genome_gene (ENTREZ_GENE_ID, GENETIC_ENTITY_ID,HUGO_GENE_SYMBOL, CYTOBAND, EXONIC_LENGTH, CHR, REFERENCE_GENOME_ID)
```
can you add a space after the comma following `GENETIC_ENTITY_ID`?

khzhu replied:
this remind me that GENETIC_ENTITY_ID is actually need to be removed.

ao508 commented on db-scripts/src/main/resources/migration.sql:
```diff
@@ -588,3 +591,15 @@ UPDATE `info` SET `DB_SCHEMA_VERSION`="2.8.1";
+    INNER JOIN `genetic_profile` ON `cancer_study`.CANCER_STUDY_ID = `genetic_profile`.CANCER_STUDY_ID
+    INNER JOIN `mutation` ON `mutation`.GENETIC_PROFILE_ID = `genetic_profile`.GENETIC_PROFILE_ID
+    INNER JOIN `mutation_event` ON `mutation`.MUTATION_EVENT_ID = `mutation_event`.MUTATION_EVENT_ID
+SET `cancer_study`.REFERENCE_GENOME_ID = IF(`mutation_event`.NCBI_BUILD in (37, 'hg19','GRCh37'), 1, 2);
```
`mutation_event.NCBI_BUILD` is `VARCHAR(10)` so shouldn't `37` be in quotes also?

khzhu replied:
right, should be. Good catch!

sheridancbio replied:
this is not yet fixed.
Also .. here we are assuming that reference genome id "1" is for hg19 and reference genome id "2" is for <anything else>. Above in the migration script we added hg19 and hg38. So this migration step may do the wrong thing if a portal installation has other builds in the database (hg18?)

khzhu replied:
quoted 37.
if you look at migration script line 399~402, reference_genome_ids are manually assigned. hg19 is always 1, hg38 is always 2. 
We only have a handful reference genomes. If anything goes wrong, we can always fix manually and update your migration.sql.

ao508 commented:
@khzhu I am reviewing this PR in phases, starting with the SQL files changed (just submitted that review). 

I will add reviews for the JAVA files separately since there are so many and probably one or two additional reviews for the remaining filetypes.

ao508 commented:
General note for the MSK pipelines team: we need to coordinate the changes to the `gene`, `reference_genome_gene` tables so that our weekly update to the database `gene` tables will remain compatible.

cc @sheridancbio @n1zea144

khzhu commented on business/src/main/resources/org/mskcc/cbio/portal/persistence/StudyMapperLegacy.xml:
```diff
@@ -14,7 +14,7 @@
         PMID as pmid,
         CITATION as citation,
-        GROUPS as groups,
+        `GROUPS` as `groups`,
```
we should always quote reserved keywords I would say.

ao508 commented on core/src/main/scripts/importer/metaImport.py:
```diff
@@ -70,9 +70,15 @@ def interface():
+    parser.add_argument('-species', '--species', type=str, default='human',
+                        help='species information (default: assumed human)',
+                        required=False)
+    parser.add_argument('-genome', '--reference_genome', type=str, default='hg19',
```
What if a user enters `-genome hg18` but does not provide an arg for `-build` - Would `build` be set to default `37` or resolve to `38`.

khzhu replied:
validateData.py script ensures the name of genome matches genome build.

ao508 commented on core/src/main/scripts/importer/validateData.py:
```diff
@@ -4025,9 +4038,15 @@ def interface(args=None):
                                    action='store_true',
                                    help='Skip tests requiring information '
                                         'from the cBioPortal installation')
-    parser.add_argument('-P', '--portal_properties', type=str,
```
okay, so looks like `portal_properties` was used exclusively for loading gene info only? (which will now be provided optionally as input arguments)

khzhu replied:
validateData script no longer uses portal.properties file.

ao508 commented on docs/Import-reference-genome.md:
```diff
@@ -0,0 +1,49 @@
+
+### What is Reference Genome (also know as reference assembly)?
+A digital nucleic acid sequence database assembled by scientists as a representative example of a species' set of genes. Reference genomes can be accessed online using 
+dedicated browsers such as [UCSC Genome Browser](https://genome.ucsc.edu/cgi-bin/hgGateway). 
```
I think your text editor must have "hard wrapped" your text. Can you fix this?

ao508 commented on docs/Import-reference-genome.md:
```diff
@@ -0,0 +1,49 @@
+
+### Why does reference genomes matter?
+Mutation and Segment plots on Patient View are using genomic coordinates. The **cytoband** column in the Mutated Gene table on Study view rely on the Genome Reference Consortium Build 
+matching the version used to analyse the original data (before it was loaded into cBioPortal DB).
```
same comment here and elsewhere, I think the rest of the file is okay

khzhu replied:
does not seem you can complete to turn hard wrap off in Intellij but you can give it a big number such as 300.

ao508 commented on docs/Using-the-dataset-validator.md:
```diff
@@ -421,25 +428,24 @@ Validation of study succeeded with warnings.
-species=human
-ncbi.build=37
-ucsc.build=hg19
+--species human
```
I don't think it hurts to have `=` signs - the arg parser should interpret the arguments the same way with or without `=` signs

ao508 replied:
(this is also not necessary)

khzhu replied:
changed.

ao508 commented on docs/Using-the-metaImport-script.md:
````diff
@@ -10,10 +10,11 @@ and then run the following command:
 This will tell you the parameters you can use:
 ```
-$ ./metaImport.py -h
+ $./metaImport.py  -h
````
leading space added should be a space after `$`

ao508 replied:
also an extra whitespace was added after `.py`

khzhu replied:
corrected.

khzhu commented on core/pom.xml:
```diff
@@ -295,7 +295,7 @@
         <configuration>
           <forkMode>always</forkMode>
-          <filtering>true</filtering>
+          <!-- filtering>true</filtering -->
```
the version mvn (3.6.0) used no longer has this tag. Not sure we still need to support old version.

ao508 commented:
@khzhu do we document anywhere which `mvn` version to use for the cbioportal?

khzhu commented on core/pom.xml:
```diff
@@ -371,15 +371,15 @@
               <!-- added encoding to prevent Unknown initial character set index '192' received from server error -->
               <encoding>UTF-8</encoding>
-              <driverProperties>characterEncoding=utf8, connectionCollation=utf8_general_ci</driverProperties>
+              <!-- driverProperties>characterEncoding=utf8, connectionCollation=utf8_general_ci</driverProperties -->
```
same above. the tag was deprecated from mvn >3.6.0.

khzhu commented on core/src/main/java/org/mskcc/cbio/portal/dao/DaoCancerStudy.java:
```diff
@@ -319,6 +319,13 @@ public static void addCancerStudy(CancerStudy cancerStudy, boolean overwrite) th
+                pstmt.setInt(12, referenceGenome.getReferenceGenomeId());
+            }
+            catch (NullPointerException e) {
+                pstmt.setInt(12,1); //#TODO default reference genome to use
```
good point! Agree, Grep reference genome information portal properties file is the best way to go.

> species and genomic information
species=human
ncbi.build=37
ucsc.build=hg19

khzhu replied:
does seems there is an easy way to read portal properties from a Dao object class. Any suggestions? 

> perhaps rather than hardcoding a "1" here, we should read a property from portal.properties with the name of a default reference genome to use when creating a study which specifies an unrecognized reference genome (or does not specify a reference genome).

khzhu replied:
@sheridancbio , are we allow to read portal.properties from a Dao object class?

khzhu replied:
@sheridancbio @ao508 , if you take look at the code block (line 108~114) in CancerStudyReader.java 
if no **_reference genome_** info included in the meta_study.txt file, CancerStudyReader actually reads from the portal.properties file.
>         String referenceGenome = properties.getProperty("reference_genome");
 So, I will remove hard coded 1 with a SQL exception to indicate an unsupported reference genome.

ao508 commented on persistence/persistence-mybatis/src/main/java/org/cbioportal/persistence/mybatis/ReferenceGenomeGeneMyBatisRepository.java:
```diff
@@ -0,0 +1,47 @@
+package org.cbioportal.persistence.mybatis;
```
cleanup extra new lines

khzhu replied:
removed.

khzhu replied:
removed extra blank line 3.

ao508 commented on persistence/persistence-mybatis/src/test/java/org/cbioportal/persistence/mybatis/PatientMyBatisRepositoryTest.java:
```diff
@@ -93,8 +93,8 @@ public void getAllPatientsInStudyDetailedProjection() throws Exception {
             " Invasive Carcinoma project. 825 cases.<br><i>Nature 2012.</i> <a href=\\\"http://tcga-data.nci." +
             "nih.gov/tcga/\\\">Raw data via the TCGA Data Portal</a>.", cancerStudy.getDescription());
         Assert.assertEquals(true, cancerStudy.getPublicStudy());
-        Assert.assertEquals("23000897,26451490", cancerStudy.getPmid());
```
multiple PMIDs are supported now.. I think these are test cases added specifically for that. Maybe this PR is out of date with the latest `rc` branch?

khzhu replied:
was the latest when this PR tested. will try again.

ao508 commented on persistence/persistence-mybatis/src/test/java/org/cbioportal/persistence/mybatis/SampleListMyBatisRepositoryTest.java:
```diff
@@ -80,8 +80,8 @@ public void getAllSampleListsDetailedProjection() throws Exception {
             " Invasive Carcinoma project. 825 cases.<br><i>Nature 2012.</i> <a href=\\\"http://tcga-data.nci." +
             "nih.gov/tcga/\\\">Raw data via the TCGA Data Portal</a>.", cancerStudy.getDescription());
         Assert.assertEquals(true, cancerStudy.getPublicStudy());
-        Assert.assertEquals("23000897,26451490", cancerStudy.getPmid());
```
same comment here as above regarding multiple PMIDs for a study

ao508 replied:
also as part of that change, the citation now ends with `, ...` if there are multiple PMIDs for a study

ao508 commented on persistence/persistence-mybatis/src/test/java/org/cbioportal/persistence/mybatis/SampleMyBatisRepositoryTest.java:
```diff
@@ -81,8 +81,8 @@ public void getAllSamplesInStudyDetailedProjection() throws Exception {
                 " Invasive Carcinoma project. 825 cases.<br><i>Nature 2012.</i> <a href=\\\"http://tcga-data.nci." +
                 "nih.gov/tcga/\\\">Raw data via the TCGA Data Portal</a>.", cancerStudy.getDescription());
         Assert.assertEquals(true, cancerStudy.getPublicStudy());
-        Assert.assertEquals("23000897,26451490", cancerStudy.getPmid());
```
same comment here

ao508 commented on persistence/persistence-mybatis/src/test/java/org/cbioportal/persistence/mybatis/StudyMyBatisRepositoryTest.java:
```diff
@@ -56,8 +56,8 @@ public void getAllStudiesSummaryProjection() throws Exception {
                 " Invasive Carcinoma project. 825 cases.<br><i>Nature 2012.</i> <a href=\\\"http://tcga-data.nci." +
                 "nih.gov/tcga/\\\">Raw data via the TCGA Data Portal</a>.", cancerStudy.getDescription());
         Assert.assertEquals(true, cancerStudy.getPublicStudy());
-        Assert.assertEquals("23000897,26451490", cancerStudy.getPmid());
```
same comment here

khzhu replied:
reverted the change.

ao508 commented on pom.xml:
```diff
@@ -294,7 +294,7 @@
   <db.test.driver>com.mysql.jdbc.Driver</db.test.driver>
     <!-- For MySQL < 5.5 change 'default_storage_engine' to 'storage_engine' -->
-  <db.test.url>jdbc:mysql://127.0.0.1:3306/cgds_test?sessionVariables=default_storage_engine=InnoDB</db.test.url>
+  <db.test.url>jdbc:mysql://127.0.0.1:3306/cgds_test</db.test.url>
```
can you switch back to `localhost`? Does the session variable need to be removed? I think I recall seeing it in a config or xml file?

khzhu replied:
127.0.0.1 is localhost. **You cannot use localhost to connect to a mysql instance running from a docker container**. 
For new version mysql jdbc driver, default storage engine is InnoDB,no need to specify using a query string.

ao508 commented on service/src/main/java/org/cbioportal/service/impl/GeneServiceImpl.java:
```diff
@@ -91,7 +100,7 @@ public Gene getGene(String geneId) throws GeneNotFoundException {
         }
 
-        chromosomeCalculator.setChromosome(gene);
+        //chromosomeCalculator.setChromosome(gene);
```
remove commented out line if not using anymore

khzhu replied:
done.

ao508 commented on service/src/main/java/org/cbioportal/service/impl/MutationServiceImpl.java:
```diff
@@ -42,7 +42,7 @@
         List<Mutation> mutationList = mutationRepository.getMutationsInMolecularProfileBySampleListId(molecularProfileId,
             sampleListId, entrezGeneIds, snpOnly, projection, pageSize, pageNumber, sortBy, direction);
```
remove commented out lines if not using them anymore

khzhu replied:
done.

ao508 commented on service/src/main/java/org/cbioportal/service/impl/ReferenceGenomeGeneServiceImpl.java:
```diff
@@ -0,0 +1,46 @@
+    
+    @Override
+    public List<ReferenceGenomeGene> fetchAllReferenceGenomeGenes(String genomeName) {
+        
```
cleanup extra new line here and below

khzhu replied:
done.

ao508 commented on service/src/main/java/org/cbioportal/service/util/AlterationEnrichmentUtil.java:
```diff
@@ -79,7 +79,7 @@ private AlterationEnrichment createAlterationEnrichment(List<? extends Alteratio
         alterationEnrichment.setEntrezGeneId(alterationCountByGene.getEntrezGeneId());
         alterationEnrichment.setHugoGeneSymbol(gene.getHugoGeneSymbol());
-        alterationEnrichment.setCytoband(gene.getCytoband());
+        //alterationEnrichment.setCytoband(gene.getCytoband()); //set value at the frontend
```
remove commented out line. If you want to capture this information then I would put it in the javadoc comment for this method

khzhu replied:
removed.

ao508 commented on service/src/main/java/org/cbioportal/service/util/ChromosomeCalculator.java:
```diff
@@ -1,6 +1,7 @@
 package org.cbioportal.service.util;
 
 import org.cbioportal.model.Gene;
+import org.cbioportal.model.ReferenceGenomeGene;
```
is this referenced anywhere here? if not then import statement should be removed

khzhu replied:
done.

ao508 commented on service/src/test/java/org/cbioportal/service/impl/MutationServiceImplTest.java:
```diff
@@ -47,16 +47,17 @@ public void getMutationsInMolecularProfileBySampleListId() throws Exception {
             SAMPLE_LIST_ID, Arrays.asList(ENTREZ_GENE_ID_1), null, PROJECTION, PAGE_SIZE, PAGE_NUMBER, SORT, DIRECTION))
             .thenReturn(expectedMutationList);
-        Mockito.doAnswer(invocationOnMock -> {
+        //TODO: varify chromosome from reference genome gene
```
verify *

ao508 commented on service/src/test/java/org/cbioportal/service/impl/MutationServiceImplTest.java:
```diff
@@ -47,16 +47,17 @@ public void getMutationsInMolecularProfileBySampleListId() throws Exception {
             .thenReturn(expectedMutationList);
-        Mockito.doAnswer(invocationOnMock -> {
+        //TODO: varify chromosome from reference genome gene
+        /*Mockito.doAnswer(invocationOnMock -> {
```
since these are `TODO`s I suppose it's okay to leave them commented here but I do believe these should be addressed in this PR

khzhu replied:
@jjgao made a comment: no need to perform the check, since is not necessary. Will remove those notes. thanks!

khzhu replied:
removed, since now we use **chr** recorded in the mutation_event table, so we do need to set a mockito chr value.

ao508 commented on service/src/test/java/org/cbioportal/service/util/AlterationEnrichmentUtilTest.java:
```diff
@@ -88,15 +89,15 @@ public void createAlterationEnrichments() throws Exception {
         Assert.assertEquals((Integer) 2, alterationEnrichment1.getEntrezGeneId());
         Assert.assertEquals("HUGO2", alterationEnrichment1.getHugoGeneSymbol());
-        Assert.assertEquals("CYTOBAND2", alterationEnrichment1.getCytoband());
+        Assert.assertEquals(null, alterationEnrichment1.getCytoband());
```
if these are just null then let's remove? if there are plans to get the cytoband from the `reference_genome_gene` then use that instead.

khzhu replied:
decided to use '-' if no cytoband found in reference_genome_gene table.

khzhu replied:
we do not support multiple genome at the profile level, genetic alterations do not carry any information on the reference genome used.  But, that is okay since we can make a API call to query reference genome genes of queried genes from the frontend to retrieve their cytobands. So, in backend the cytoband of any alterations has a null value.

khzhu commented on core/src/main/scripts/importer/cbio_importer.py:
```diff
@@ -0,0 +1,185 @@
+import validateData
+
+def send_mail(physician, patient, sample):
+    me = "cbioportal@uhnresearch.ca"
```
that is a side effect of **_git add -A_**, the script is meant to be an internal use, sorry, will remove.

khzhu commented on core/src/main/scripts/importer/cbio_importer.py:
```diff
@@ -0,0 +1,185 @@
+import sys
```
cbio_importer.py file will be excluded from the PR.

khzhu commented on core/src/main/java/org/mskcc/cbio/portal/dao/DaoGeneOptimized.java:
```diff
@@ -362,7 +362,8 @@ public CanonicalGene getGeneByEntityId(int geneticEntityId) {
             List<CanonicalGene> ret = new ArrayList<CanonicalGene>();
             for (CanonicalGene cg : genes) {
-                String gchr = getChrFromCytoband(cg.getCytoband());
+                //String gchr = getChrFromCytoband(cg.getCytoband());
```
a very good point! I search the project this **_chr_** param no longer in use. Will remove chr from the function and update the JavaDoc.

khzhu commented on core/src/main/java/org/mskcc/cbio/portal/dao/DaoReferenceGenome.java:
```diff
@@ -0,0 +1,299 @@
+/*
```
will correct those formats, might be my intelliJ not set up right.

khzhu replied:
done.

khzhu commented on core/src/main/java/org/mskcc/cbio/portal/dao/DaoReferenceGenome.java:
```diff
@@ -0,0 +1,299 @@
+    private static final Map<String, Integer> genomeInternalIds = new HashMap<String,Integer>();
+
+    static {
+        SpringUtil.initDataSource();
```
sure not be a problem, since SpringUtil class only initialize the data source if context is null.

khzhu commented on core/src/main/java/org/mskcc/cbio/portal/dao/DaoReferenceGenome.java:
```diff
@@ -0,0 +1,299 @@
+     */
+    public static void addReferenceGenome(ReferenceGenome referenceGenome) throws DaoException {
+        reCache();
+        addReferenceGenome(referenceGenome, false);
```
it is okay. please see Robert's comments.

khzhu commented on core/src/main/java/org/mskcc/cbio/portal/dao/DaoReferenceGenome.java:
```diff
@@ -0,0 +1,299 @@
+
+                pstmt.executeUpdate();
+                rs = pstmt.getGeneratedKeys();
+                if (rs.next()) {
```
good catch! Will check the return value.

khzhu commented on core/src/main/java/org/mskcc/cbio/portal/dao/DaoReferenceGenome.java:
```diff
@@ -0,0 +1,299 @@
+            rows += pstmt.executeUpdate();
+            // update reference cache
+            reCache();
+            return rows;
```
very good point! changed to void.

khzhu commented on core/src/main/java/org/mskcc/cbio/portal/dao/DaoReferenceGenome.java:
```diff
@@ -0,0 +1,299 @@
+     */
+
+    public static ReferenceGenome getReferenceGenomeByGenomeName(String genomeName) throws DaoException {
+        //return byGenomeName.get(genomeName);
```
sorry, did some testing and forgot to remove the blocks. Should grep reference genome of interest from the cache.

khzhu commented on core/src/main/java/org/mskcc/cbio/portal/dao/DaoReferenceGenome.java:
```diff
@@ -0,0 +1,299 @@
+            pstmt = con.prepareStatement("SELECT * FROM reference_genome WHERE `name` = ?");
+            pstmt.setString(1,genomeName);
+            rs = pstmt.executeQuery();
+            if (rs.next()) {
```
a good question. In theory, we can, but we (JJ, Benjamin, Pieter) decided to make our lives easier, to give a different genome build a different genome name. hg19 and hg38 essentially are two genome assemblies for human.
agree, will add a unique key for genomeName.

khzhu commented on core/src/main/java/org/mskcc/cbio/portal/dao/DaoReferenceGenome.java:
```diff
@@ -0,0 +1,299 @@
+            e.printStackTrace();
+        } finally {
+            JdbcUtil.closeAll(DaoReferenceGenome.class, con, pstmt, rs);
+            return referenceGenome;
```
good catch, definitely should not, my mistake, will make a change.

khzhu commented on core/src/main/java/org/mskcc/cbio/portal/dao/DaoReferenceGenome.java:
```diff
@@ -0,0 +1,299 @@
+     * @param name   Name of Reference Genome or Genome Assembly
+     * @throws DaoException Database Error.
+     */
+    @Deprecated
```
all right, will remove it.

khzhu replied:
gave it a second thought. We might want to remove deprecated tag, since we already decided we are not going to support multiple species per portal installation, and most portal will be for human. Let's keep it, I would say.

khzhu commented on core/src/main/java/org/mskcc/cbio/portal/dao/DaoReferenceGenome.java:
```diff
@@ -0,0 +1,299 @@
+     * @param species genetic species               
+     * @throws DaoException Database Error.
+     */
+    public static int getReferenceGenomeIdByName(String name, String species) throws DaoException {
```
good point. will remove species.

khzhu commented on core/src/main/java/org/mskcc/cbio/portal/dao/DaoReferenceGenomeGene.java:
```diff
@@ -0,0 +1,255 @@
+                rows += pstmt.executeUpdate();
+            }
+            else {
+                rows += updateGene(existingGene);
```
I do collect multiple updates for a group of genes.

khzhu replied:
change to void function.

khzhu commented on core/src/main/java/org/mskcc/cbio/portal/dao/DaoReferenceGenomeGene.java:
```diff
@@ -0,0 +1,255 @@
+        //Exists only for default instantiation
+    }
+
+    public static DaoReferenceGenomeGene getInstance() {
```
okay, will remove default member constructor.

khzhu commented on core/src/main/java/org/mskcc/cbio/portal/dao/DaoReferenceGenome.java:
```diff
@@ -0,0 +1,299 @@
+    public static int getReferenceGenomeIdByName(String name, String species) throws DaoException {
+        try {
+            return genomeInternalIds.get(name);
+        } catch (java.lang.NullPointerException exp) {
```
agree. Updated the logic. thanks!

khzhu commented on core/src/main/java/org/mskcc/cbio/portal/dao/DaoReferenceGenome.java:
```diff
@@ -0,0 +1,299 @@
+            } else if (species.equals(ReferenceGenome.MUS_MUSCULUS))  {
+                return genomeInternalIds.get(ReferenceGenome.MUS_MUSCULUS_DEFAULT_GENOME_BUILD); // NCBI_BUILD field was an optional in the past
+            } else {
+                throw new DaoException("Species not supproted yet");
```
species removed. We will not need to support query by the specie for now :).

khzhu commented on core/src/main/java/org/mskcc/cbio/portal/dao/DaoReferenceGenomeGene.java:
```diff
@@ -0,0 +1,255 @@
+        if (instance == null) {
+            instance = new DaoReferenceGenomeGene();
+        }
+        return  instance;
```
removed.

khzhu commented on core/src/main/java/org/mskcc/cbio/portal/dao/DaoReferenceGenomeGene.java:
```diff
@@ -0,0 +1,255 @@
+        Connection con = null;
+        PreparedStatement pstmt = null;
+        ResultSet rs = null;
+        boolean setBulkLoadAtEnd = false;
```
removed.

khzhu commented on core/src/main/java/org/mskcc/cbio/portal/dao/DaoReferenceGenomeGene.java:
```diff
@@ -0,0 +1,255 @@
+            pstmt.setLong(6, gene.getEntrezGeneId());
+            pstmt.setInt(7, gene.getReferenceGenomeId());
+
+            rows += pstmt.executeUpdate();
```
is the same, removed +.

khzhu commented on core/src/main/java/org/mskcc/cbio/portal/dao/DaoReferenceGenomeGene.java:
```diff
@@ -0,0 +1,255 @@
+                ProgressMonitor.logWarning("No change for " + gene.getEntrezGeneId() + " " + gene.getReferenceGenomeId() + "? Code " + rows);
+            }
+
+            return rows;
```
I calls this function from importGeneData for a list of genes. So, I prefer to have 0 or 1 returned to know how many rows in total be updated.

khzhu replied:
agree. changed to void.

khzhu commented on core/src/main/java/org/mskcc/cbio/portal/dao/DaoReferenceGenome.java:
```diff
@@ -0,0 +1,299 @@
+ */
+public final class DaoReferenceGenome {
+
+    private static final Map<String,ReferenceGenome> byGenomeBuild = new HashMap<String,ReferenceGenome>();
```
Let's revisit here if performance become an issue.

khzhu commented on core/src/main/java/org/mskcc/cbio/portal/dao/DaoReferenceGenomeGene.java:
```diff
@@ -0,0 +1,255 @@
+                pstmt.setInt(5, gene.getExonicLength());
+                pstmt.setLong(6, gene.getStart());
+                pstmt.setLong(7, gene.getEnd());
+                rows += pstmt.executeUpdate();
```
changed to =.

khzhu commented on core/src/main/java/org/mskcc/cbio/portal/scripts/ImportExtendedMutationData.java:
```diff
@@ -146,7 +146,18 @@ public void importData() throws IOException, DaoException {
+                }
 
+                if (!record.getNcbiBuild().equalsIgnoreCase(genomeBuildName)) {
+                    ProgressMonitor.setCurrentMessage("Genome Build Name does not match, expecting " + genomeBuildName);
```
NCBI Build name currently is an option field for the maf file. Was agreed at meeting with JJ, Pieter, Benjamin, the default genome will be used if no build name supplied.
Will read the reference genome information from the portal.properties file not using a symbolic link.

khzhu replied:
genome build names never be normalized (hg19, GRCh37, 37), so we only send out a warning messages not raise an exception or stop importing.

pieterlukasse replied:
There are 2 scenarios here: (1) if no build info, fall back to default: not ideal, but OK (I don't see it implemented though); (2) build info is wrong: please throw an error in this case.

khzhu replied:
@pieterlukasse , it was agreed during meeting with @jjgao and @n1zea144, we only send out a warning message.

pieterlukasse replied:
Ok, given it is already strictly checked for `human` and `mouse` at https://github.com/cBioPortal/cbioportal/blob/83f4c918e12ec83121a1821a42044abcd33fe4ab/core/src/main/scripts/importer/validateData.py#L1686 (`checkNCBIbuild(self, value)`), resulting in validation error, I can live with no real check in this layer. Once the validation is extended to other species (perhaps rat) or restricted to only human and mouse, then this code can actually be removed from the java layer.
Only remark for now is: use `ProgressMonitor.logWarning` instead of `ProgressMonitor. setCurrentMessage`

khzhu replied:
changed to logWarning.

khzhu commented on core/src/main/java/org/mskcc/cbio/portal/scripts/ImportReferenceGenome.java:
```diff
@@ -0,0 +1,184 @@
+                String releaseDate = parts[5];
+
+                ReferenceGenome referenceGenome = new ReferenceGenome(name, species, buildName);
+                if (StringUtils.isNotEmpty(url)) {
```
I believe isNotNullOrEmpty has been deprecated, isNotEmpty() does the same thing.

khzhu commented on core/src/main/java/org/mskcc/cbio/portal/scripts/ImportReferenceGenome.java:
```diff
@@ -0,0 +1,184 @@
+     * @throws DaoException
+     */
+    private static void addReferenceGenomesToDB(Set<ReferenceGenome> referenceGenomes) throws DaoException {
+
```
removed.

khzhu replied:
removed.

khzhu commented on core/src/main/java/org/mskcc/cbio/portal/servlet/GetCoExpressionJSON.java:
```diff
@@ -242,8 +241,7 @@ protected void doPost(HttpServletRequest httpServletRequest,
                                 fullResutlStr.append(
-                                    comparedGene.getHugoGeneSymbolAllCaps() + "\t" +
-                                    comparedGene.getCytoband() + "\t" +
+                                    comparedGene.getHugoGeneSymbolAllCaps() + "\t" + "\t" +
```
nice catch, removed the extra tab.

khzhu commented on core/src/main/scripts/importer/cbio_importer.py:
```diff
@@ -0,0 +1,185 @@
+        sample_list_new = get_sample_info(connection)
+        new_samples = get_sample_dic(sample_list_new)
+        logger.info(new_samples)
+    # try import study
```
thanks for reviewing the extra code, should not be part of the PR.

khzhu commented on core/src/test/java/org/mskcc/cbio/portal/scripts/TestImportGeneData.java:
```diff
@@ -65,11 +68,11 @@ public void testImportGeneData() throws Exception {
         File file = new File("src/test/resources/supp-genes.txt");
-
-        ImportGeneData.importSuppGeneData(file);
+        ImportGeneData.importSuppGeneData(file, ReferenceGenome.HOMO_SAPIENS_DEFAULT_GENOME_BUILD);*/
```
agree, will remove.

khzhu commented on core/src/test/java/org/mskcc/cbio/portal/scripts/TestImportGeneData.java:
```diff
@@ -96,25 +99,31 @@ public void testImportGeneData() throws Exception {
         CanonicalGene gene = daoGene.getNonAmbiguousGene("ABCA4", "chr1", false);
-        assertEquals(372,gene.getLength());
+        refGene = DaoReferenceGenomeGene.getInstance().getGene("ABCA4",1);
+        //assertEquals(372,gene.getLength());
```
removed.

khzhu commented on core/src/test/resources/applicationContext-dao.xml:
```diff
@@ -47,7 +47,7 @@
 	<bean id="dbcpDataSource" destroy-method="close" class="org.apache.commons.dbcp2.BasicDataSource">
 		<property name="driverClassName" value="com.mysql.jdbc.Driver" />
-		<property name="url" value="jdbc:mysql://localhost:3306/cgds_test" />
+		<property name="url" value="jdbc:mysql://127.0.0.1:3306/cgds_test" />
```
127.0.0.1 is just your localhost. You can not use localhost if you were running mysql from a docker container.

khzhu commented on model/src/main/java/org/cbioportal/model/ReferenceGenome.java:
```diff
@@ -0,0 +1,176 @@
+    private String genomeName;
+    private String species;
+    private String buildName; //genome assembly name
+    private long genomeSize; //non-N bases
```
for myBatis ORM mapper.

khzhu replied:
The plan was to get genomic coordinates from an API. Currently, we hard-cod the genomic coordinates in a JSON file.

khzhu commented on model/src/main/java/org/cbioportal/model/ReferenceGenomeGene.java:
```diff
@@ -0,0 +1,90 @@
+    private Long start;
+    private Long end;
+    
+    public void setReferenceGenomeId(Integer referenceGenomeId) { this.referenceGenomeId = referenceGenomeId; }
```
you mean with/without this?

khzhu replied:
Keep one line style like Mutation. I still see mixed blocks style and one line style in our model classes though.

khzhu commented:
thanks @sheridancbio and @ao508 very much for taking time review the PR. I addressed most of issues raised by your code reviews. There were a few places that I did not go with your suggestions, I either gave my explanation or proposed a different approach. Please let me know if anything I missed should be addressed in this PR to get your approval. Thank you!

khzhu commented on core/src/main/java/org/mskcc/cbio/portal/dao/DaoCancerStudy.java:
```diff
@@ -624,6 +631,12 @@ private static CancerStudy extractCancerStudy(ResultSet rs) throws SQLException
+            cancerStudy.setReferenceGenome(DaoReferenceGenome.getReferenceGenomeByInternalId(
+                rs.getInt("REFERENCE_GENOME_ID")).getGenomeName());
+        } catch (DaoException e) {
+            cancerStudy.setReferenceGenome(ReferenceGenome.HOMO_SAPIENS_DEFAULT_GENOME_NAME);
```
let's have the default reference genome name/build in portal.properties.

averyniceday replied:
@khzhu is this change going to be put into the current PR? just checking, thanks!

khzhu commented on core/src/main/java/org/mskcc/cbio/portal/model/ReferenceGenomeGene.java:
```diff
@@ -0,0 +1,129 @@
+    }
+    
+
+    public void setReferenceGenomeId(int referenceGenomeId) { this.referenceGenomeId = referenceGenomeId; }
```
is a common practice, not have **_this_** for getter.

khzhu replied:
changed all to a multi-line block.

khzhu commented on persistence/persistence-api/src/main/java/org/cbioportal/persistence/ReferenceGenomeGeneRepository.java:
```diff
@@ -0,0 +1,43 @@
+/*
+ * Copyright (c) 2016 Memorial Sloan Kettering Cancer Center.
```
you are right, should use 2019. Will update.

khzhu replied:
dropped the header.

onursumer commented on web/src/main/java/org/cbioportal/web/ReferenceGenomeGeneController.java:
```diff
@@ -0,0 +1,86 @@
+    @RequestMapping(value = "/reference-genome-genes/{genomeName}/fetch", method = RequestMethod.POST, consumes = MediaType.APPLICATION_JSON_VALUE,
+        produces = MediaType.APPLICATION_JSON_VALUE)
+    @ApiOperation("Fetch genes of reference genome of interest")
+    public ResponseEntity<List<ReferenceGenomeGene>> fetchGenes(
```
Can we rename this method to `fetchReferenceGenomeGenes`? We have another `fetchGenes` method in `GeneController` and when you auto generate API client, you get function names like `fetchGenesUsingPOST_1` due to name conflicts.

When implementing controllers, try to match the name of your functions to the name of your model.

khzhu replied:
@onursumer, I see. change made, thanks for the tip.

khzhu commented on web/src/main/java/org/cbioportal/web/ReferenceGenomeGeneController.java:
```diff
@@ -0,0 +1,86 @@
+
+    @RequestMapping(value = "/reference-genome-genes/{genomeName}", method = RequestMethod.GET, produces = MediaType.APPLICATION_JSON_VALUE)
+    @ApiOperation("Get all reference genes")
+    public ResponseEntity<List<ReferenceGenomeGene>> getAllRefGenes(
```
change made.

khzhu commented on web/src/main/java/org/cbioportal/web/ReferenceGenomeGeneController.java:
```diff
@@ -0,0 +1,86 @@
+
+    @RequestMapping(value = "/reference-genome-genes/{genomeName}/{geneId}", method = RequestMethod.GET, produces = MediaType.APPLICATION_JSON_VALUE)
+    @ApiOperation("Get a gene of a reference genome of interest")
+    public ResponseEntity<ReferenceGenomeGene> getGene(
```
change made.

averyniceday commented on core/src/main/java/org/mskcc/cbio/portal/dao/DaoGeneOptimized.java:
```diff
@@ -436,19 +407,18 @@ public CanonicalGene getNonAmbiguousGene(String geneId) {
      * @return a gene that can be non-ambiguously determined, or null if cannot.
      */
     public CanonicalGene getNonAmbiguousGene(String geneId, String chr) {
```
Looks like chr argument can be dropped here.

khzhu replied:
good catch! will remove chr.

khzhu replied:
chr param removed.

averyniceday commented on web/src/main/java/org/cbioportal/web/util/ClinicalDataEnrichmentUtil.java:
```diff
@@ -19,7 +19,7 @@
 import org.springframework.beans.factory.annotation.Autowired;
 import org.springframework.stereotype.Component;
-
+ 
```
whitespace introduced here

khzhu replied:
removed.

khzhu commented:
as of today, all review comments have been resolved and the PR passed CI. Please let me know anything need to be done to get an approval. Thank you!

sheridancbio commented:
@n1zea144 @jjgao @pieterlukasse : we are getting close to completing the review of this PR and hopefully merging this work soon. According to the new approach in CONTRIBUTING.md, I will soon be closing the outstanding requests for reviews to you, unless you want to do so yourself or let me know today that you are planning to submit a review.
@onursumer @ao508 : If changes made by @khzhu seem to address the changes requested in your reviews of this PR, you can approve this PR if you like. Alternatively, I can approve the overall PR after I complete my review and can mark your change requests as being satisfied if they do seem to be satisfied. (Dismissing the change requests due to fulfillment)

pieterlukasse commented:
@sheridancbio thanks for the heads up. I will take a look at it tomorrow.

sheridancbio commented on docs/Using-the-dataset-validator.md:
```diff
@@ -421,25 +428,24 @@ Validation of study succeeded with warnings.
-ucsc.build=mm10
+--species=mouse
+--reference_genome=mm10
+--genome_buil=38
```
typo: --genome_buil

khzhu replied:
good catch! change made. thanks!

sheridancbio commented:
A general comment : I'm looking into the question of using portal.properties as the definition of the default reference genome to assume during operations. It seems as though code to parse portal.properties was added to the core module, but that there are still hardcoded default options in various classes (e.g. DaoCancerStudy.java contains a reference to the hardcoded ReferenceGenome.HOMO_SAPIENS_DEFAULT_GENOME_NAME). There are a few places in the PR to check regarding this. I also see that the validator script was altered to not read the portal.properties file and to instead take additional command line arguments which override hardcoded defaults. Maybe this is ok ... especially if we plan to always provide the command line arguments to the validator. But I will probably be suggesting soon that the core module not have embedded choices for default genomes to use when they are not specified. Externalizing those choices to the properties files makes them more visible for when it is time to jettison the core module.

khzhu commented:
@sheridancbio ,

>  It seems as though code to parse portal.properties was added to the core module, but that there are still hardcoded default options in various classes (e.g. DaoCancerStudy.java contains a reference to the hardcoded ReferenceGenome.HOMO_SAPIENS_DEFAULT_GENOME_NAME).
good catch. daoException should be thrown instead. 

nice catch! StudyReader reads **_reference_genome_** property from meta_study.txt file, if it could not find the property in the file, it will use the default genome specified in portal.properties file. So, I should raise a DaoException instead.

> I also see that the validator script was altered to not read the portal.properties file and to instead take additional command line arguments which override hardcoded defaults. 

yes, that is a desired behavior. since we support multiple genome buils. portal.properties file can only specify one reference genome. Both -genome and -build are optional parameter to the validator scripts. If users do not supply those two options, the default genome build listed in the portal.properties file will be used.

sheridancbio commented on web/src/main/java/org/cbioportal/web/ReferenceGenomeGeneController.java:
```diff
@@ -0,0 +1,86 @@
+                referenceGenomeGeneService.fetchAllReferenceGenomeGenes(genomeName), HttpStatus.OK);
+    }
+
+    @RequestMapping(value = "/reference-genome-genes/{genomeName}/{geneId}", method = RequestMethod.GET, produces = MediaType.APPLICATION_JSON_VALUE)
```
This is good .. the endpoint query arguments has genomeName strings rather than referenceGenomeId (integers). Also Entrez_gene_id is good (rather than the internal genetic_entity_id). The one thing I am concerned with here is that I think we are exposing the internal id for the referenceGenome in the output from the API call. referenceGenomeId is a data member of ReferenceGenomeGene so it is being serialized in the output here. I would recommend that we inhibit the output of this auto-increment field when serializing this data in the output to the web APIs. I believe we already do that already for other data types such as Sample through the CustomObjectMapper class using the "mixin" technique. You can see in SampleMixin.java there is an annotation "@JsonIgnore" over the internalId member in the Sample model. That prevents our internal identifiers from being exposed to outside users. I think I request that we do the same thing for the referenceGenomeId in the ReferenceGenomeGene model.

khzhu replied:
very good points! Definitely, I will add one for ReferenceGenomeGene.

khzhu replied:
added JsonIgnore for referenceGenomeId.

khzhu commented on core/src/main/java/org/mskcc/cbio/portal/dao/DaoReferenceGenome.java:
```diff
@@ -0,0 +1,259 @@
+     * @throws DaoException Database Error.
+     */
+
+    public static ReferenceGenome getReferenceGenomeByBuildName(String buildName) throws DaoException {
```
@sheridancbio , let's keep it for future use. Query reference genomes by the build name seems to be more correct to me.

khzhu replied:
will add Minix class. Thanks for the explanation!

khzhu commented on core/src/main/java/org/mskcc/cbio/portal/dao/DaoReferenceGenomeGene.java:
```diff
@@ -0,0 +1,177 @@
+        PreparedStatement pstmt = null;
+        ResultSet rs = null;
+        try {
+            ReferenceGenomeGene existingGene = null;
```
agree, will change.

khzhu replied:
change made.

sheridancbio commented on db-scripts/src/main/resources/migration.sql:
```diff
@@ -599,3 +590,24 @@ CREATE TABLE `data_access_tokens` (
+UPDATE `cancer_study`
+    INNER JOIN `genetic_profile` ON `cancer_study`.CANCER_STUDY_ID = `genetic_profile`.CANCER_STUDY_ID
+    INNER JOIN `mutation` ON `mutation`.GENETIC_PROFILE_ID = `genetic_profile`.GENETIC_PROFILE_ID
+    INNER JOIN `mutation_event` ON `mutation`.MUTATION_EVENT_ID = `mutation_event`.MUTATION_EVENT_ID
```
this is a big join ... It may take a while to run. Maybe we should test this script on our production database and see if it completes.

khzhu replied:
please test out. it worked for us.

sheridancbio commented on web/src/test/java/org/cbioportal/web/ReferenceGenomeGeneControllerTest.java:
```diff
@@ -0,0 +1,141 @@
+            .andExpect(MockMvcResultMatchers.content().contentTypeCompatibleWith(MediaType.APPLICATION_JSON))
+            .andExpect(MockMvcResultMatchers.jsonPath("$.entrezGeneId").value(ENTREZ_GENE_ID_1))
+            .andExpect(MockMvcResultMatchers.jsonPath("$.referenceGenomeId").value(REFERENCE_GENOME_ID))
+            .andExpect(MockMvcResultMatchers.jsonPath("$.cytoband").value(CYTOBAND_1))
```
here we are testing that the cytoband differs depending on the reference-genome-gene .. but we are testing that it differs in our mocks, not through the service/persistence layer business logic. So the tests here are fine ... but we still might want some tests at a lower layer showing that the returned cytoband does come from the linked reference_genome_gene (if we don't yet have it)

sheridancbio commented:
Canceling review request from @jjgao and @n1zea144 now

pieterlukasse commented on db-scripts/src/main/resources/migration.sql:
```diff
@@ -599,3 +590,24 @@ CREATE TABLE `data_access_tokens` (
+ALTER TABLE `gene` DROP COLUMN `CYTOBAND`, DROP COLUMN `LENGTH`;
+ALTER TABLE `cancer_study` ADD COLUMN `REFERENCE_GENOME_ID` INT(4) DEFAULT 1,
+                           ADD CONSTRAINT `FK_REFERENCE_GENOME` FOREIGN KEY (`REFERENCE_GENOME_ID`)
+                               REFERENCES `reference_genome`(`REFERENCE_GENOME_ID`) ON DELETE CASCADE;
```
in `cgds.sql` this is `ON DELETE RESTRICT` and here it is `ON DELETE CASCADE`. Please adjust to make them the same. My preference goes to `RESTRICT` in this case. 
As a general comment: please put the cgds.sql and migration.sql side by side and check very carefully to make sure both scripts end up with the same result.

khzhu replied:
what do you think @sheridancbio , should we go with RESTRICT? I vote for ON DELETE CASCADE.

pieterlukasse replied:
any updates here?

khzhu replied:
what do you think, @sheridancbio? We may never want to remove any reference genomes. Either way really does not matter I would say.

khzhu replied:
@pieterlukasse , changed to **_ON DELETE RESTRICT_** to be consist with cgds.sql.

pieterlukasse commented on db-scripts/src/main/resources/cgds.sql:
```diff
@@ -793,7 +793,7 @@ CREATE TABLE `clinical_event_data` (
     `ENTREZ_GENE_ID` int(11) NOT NULL,
     `REFERENCE_GENOME_ID` int(4) NOT NULL,
-    `CHR` varchar(4) DEFAULT NULL,
+    `CHR` varchar(5) DEFAULT NULL,
```
I didn't find this change in `migration.sql`

khzhu replied:
good catch! will update migration.sql.

khzhu replied:
updated.

sheridancbio commented on core/src/main/java/org/mskcc/cbio/portal/dao/DaoReferenceGenome.java:
```diff
@@ -0,0 +1,259 @@
+                pstmt.setLong(4, referenceGenome.getGenomeSize());
+                pstmt.setString(5, referenceGenome.getUrl());
+                pstmt.setDate(6, new java.sql.Date(referenceGenome.getReleaseDate().getTime()));
+                if (pstmt.executeUpdate() != -1 ) {
```
the return value from executeUpdate is never -1 .. it will be 0 if no records were altered, or it will be a record count if some records were altered. So I think this statement should be 
if (pstmt.executeUpdate() != 1 )
Also, please add an "else" clause to this statement which something like:
else {
throw new DaoException("attempt to add new referenceGenome record failed")
}

khzhu replied:
good catch! Will make changes.

khzhu replied:
changed. Progress monitor indicates update statement that affected zero rows.
no need for an else statement since try..catch block will handle any exceptions.
```
            if (pstmt.executeUpdate() == 0) {
                ProgressMonitor.logWarning("No change for " + gene.getEntrezGeneId() + " " + gene.getReferenceGenomeId());
            }
```

khzhu commented on docs/Using-the-dataset-validator.md:
```diff
@@ -38,8 +38,12 @@ optional arguments:
+  -genome REFERENCE_GENOME, --reference_genome REFERENCE_GENOME
+                        reference genome build (default: assumed hg19)
+  -build GENOME_BUILD, --genome_build GENOME_BUILD
+                        reference genome build (default: assumed 37 for reference genome hg19)
```
we should allow to have multiple genome builds for the same genome. They are totally different concepts. If we need to choose one or other, I would vote for use **_genome build_**.

pieterlukasse replied:
but the combination genome=`hg19` and build=`38` is wrong, right?

khzhu replied:
the validator script verifies the combination of genome name and build.

khzhu replied:
@pieterlukasse , I would suggest keep the way it is now till we humanized those vocabularies in the mutation_event table. And, then we complete get rid of the name of genome as well. Let's leave this for a future PR.

khzhu replied:
@pieterlukasse , agree, changed to 
```
--ucsc_build_name
--ncbi_build_name
```
since --grch_build_number does not work for mouse. also users are now only required to supply either one.

khzhu commented on service/src/test/java/org/cbioportal/service/impl/DiscreteCopyNumberServiceImplTest.java:
```diff
@@ -1,11 +1,6 @@
-import org.cbioportal.model.Gene;
-import org.cbioportal.model.GeneMolecularData;
-import org.cbioportal.model.MolecularProfile;
+import org.cbioportal.model.*;
```
will do that. thanks!

khzhu replied:
done.

khzhu commented on docs/Using-the-dataset-validator.md:
```diff
@@ -421,25 +428,24 @@ Validation of study succeeded with warnings.
-ucsc.build=mm10
+--species=mouse
+--reference_genome=mm10
+--genome_build=38
```
mm10 is UCSC mouse genome assembly name, GRCm38 is NCBI mouse build name. But, I found most time people use mm10 to reference mouse genome.

khzhu commented on db-scripts/src/main/resources/cgds.sql:
```diff
@@ -112,6 +111,20 @@ CREATE TABLE `type_of_cancer` (
+    `RELEASE_DATE` datetime DEFAULT NULL,
+    PRIMARY KEY (`REFERENCE_GENOME_ID`),
+    UNIQUE INDEX `BUILD_NAME_UNIQUE` (`BUILD_NAME` ASC),
+    CHECK(`SPECIES` = 'human')
```
good catch! will remove the check. The importer script does check the species agains the values listed in the portal.properties file.

khzhu replied:
species constraint check removed.

pieterlukasse replied:
Ok, check removed. But I added some comments regarding validation part.

khzhu commented on core/src/main/java/org/mskcc/cbio/portal/model/ReferenceGenomeGene.java:
```diff
@@ -0,0 +1,129 @@
+ */
+
+package org.mskcc.cbio.portal.model;
+
```
removed.

khzhu commented on core/src/main/java/org/mskcc/cbio/portal/scripts/ImportExtendedMutationData.java:
```diff
@@ -146,7 +146,18 @@ public void importData() throws IOException, DaoException {
+                    String referenceGenome = cancerStudy.getReferenceGenome();
+                    genomeBuildName = DaoReferenceGenome.getReferenceGenomeByGenomeName(referenceGenome).getBuildName();
+                } catch (NullPointerException e) {
+                    genomeBuildName = GlobalProperties.DEFAULT_UCSC_BUILD;
```
very good point! added a new static method to GlobalProperties class.
```
    public static String getReferenceGenomeBuild() {
        return portalProperties.getProperty(UCSC_BUILD, DEFAULT_UCSC_BUILD);
    }
```

khzhu commented on core/src/main/java/org/mskcc/cbio/portal/scripts/ImportPathwayCommonsExtSif.java:
```diff
@@ -82,10 +82,10 @@ public void importData() throws IOException, DaoException {
             String geneAId = parts[0];
 
-            CanonicalGene geneA = daoGene.getNonAmbiguousGene(geneAId, null);
+            CanonicalGene geneA = daoGene.getNonAmbiguousGene(geneAId);
```
default value for second param is "true".

ao508 replied:
@khzhu Just wanted to clarify a bit more - the call `daoGene.getNonAmbiguousGene(geneAId, null);` --> `daoGene.getNonAmbiguousGene(geneAId)` isn't consistent with other calls that were modified in this pr, such as the one [below at L88](https://github.com/cBioPortal/cbioportal/pull/5891/files#diff-234c49aafd25c753e7488792fbc4762aR88). Why was  the second parameter ignored above but set to `true` in the call below?

khzhu replied:
okay added in to be consistent.

khzhu commented on core/src/main/java/org/mskcc/cbio/portal/util/CancerStudyReader.java:
```diff
@@ -104,6 +105,11 @@ private static CancerStudy getCancerStudy(TrimmedProperties properties)
         cancerStudy.setShortName(shortName);
+        String referenceGenome = properties.getProperty("reference_genome");
+        if (referenceGenome == null) {
+            referenceGenome = ReferenceGenome.HOMO_SAPIENS_DEFAULT_GENOME_NAME;
```
changed. Now, get default reference genome from portal.properties file.
```
if (referenceGenome == null) {
            referenceGenome = GlobalProperties.getReferenceGenomeBuild();
}
```

khzhu commented on core/src/main/java/org/mskcc/cbio/portal/util/EnrichmentsAnalysisUtil.java:
```diff
@@ -642,4 +652,20 @@ private double runFisherExactTest(HashMap<Integer, String> singleGeneCaseValueMa
+            try {
+                genomeName = DaoCancerStudy.getCancerStudyByInternalId(cancerStudyId).getReferenceGenome();
+            } catch (NullPointerException ne) {
+                genomeName = ReferenceGenome.HOMO_SAPIENS_DEFAULT_GENOME_NAME;
```
changed. Now, get from portal.properties file.

khzhu commented on core/src/main/scripts/importer/validateData.py:
```diff
@@ -3737,29 +3750,29 @@ def validate_defined_caselists(cancer_study_id, case_list_ids, file_types, logge
-                "case list to specify which samples are profiled for this data type. On the query page, this "
-                "case list will be selected by default when both mutation and CNA data are available.",
-                cancer_study_id + '_cnaseq')
+        if 'meta_mutations_extended' in file_types:
```
done.

khzhu commented on core/src/test/java/org/mskcc/cbio/portal/scripts/TestImportReferenceGenome.java:
```diff
@@ -0,0 +1,79 @@
+        File file = new File("src/test/resources/reference_genomes.txt");
+
+        ImportReferenceGenome.importData(file);
+        
```
removed.

khzhu commented on core/src/test/resources/reference_genomes.txt:
```diff
@@ -0,0 +1,4 @@
+#species	name	build_name	nonN_bases	URL	release_date
+#human	hg19	GRCh37	2897310462	http://hgdownload.cse.ucsc.edu/goldenPath/hg19	2009-02-01 00:00:00
```
line should start with "#". uncomment the line 2 and 3.

khzhu commented on db-scripts/src/main/resources/migration.sql:
```diff
@@ -599,3 +590,24 @@ CREATE TABLE `data_access_tokens` (
+    ENTREZ_GENE_ID,
+    CYTOBAND,
+    LENGTH,
+    SUBSTRING_INDEX(SUBSTRING_INDEX(SUBSTRING_INDEX(gene.CYTOBAND,IF(LOCATE('p', gene.CYTOBAND), 'p', 'q'), 1),'q',1),'cen',1),
```
tested and worked well.

khzhu commented on docs/File-Formats.md:
```diff
@@ -620,7 +620,7 @@ The extended MAF format recognized by the portal has:
 2. **Entrez_Gene_Id (Optional, but recommended)**: A [Entrez Gene](http://www.ncbi.nlm.nih.gov/gene) identifier.
 3. **Center (Optional)**: The sequencing center.
-4. **NCBI_Build (Optional)<sup>1</sup>**: Must be "GRCh37" for human, and "GRCm38" for mouse.
+4. **NCBI_Build (Required)<sup>1</sup>**: The Genome Reference Consortium Build is used by a variant calling software. It must be "GRCh37" or "GRCh38" for a human, and "GRCm38" for a mouse.
```
it was agreed we only support hg19/hg38. mm9/mm10 at the moment. We can always update document if we need to support a new genome.

ao508 replied:
@khzhu we seem to refer to the NCBI builds as `GRCh*` but the python scripts just refer (or expect) the actual numbers. These should match. I think the python scripts should be changed to follow `GRCh*` for consistency

khzhu commented on model/src/main/java/org/cbioportal/model/CancerStudy.java:
```diff
@@ -30,6 +30,7 @@
     private Integer methylationHm27SampleCount;
     private Integer rppaSampleCount;
     private Integer completeSampleCount;
+    private String referenceGenome;
```
very good point! Added JSON property for reference genome.

khzhu commented on model/src/main/java/org/cbioportal/model/CancerStudy.java:
```diff
@@ -214,4 +215,8 @@ public Integer getCompleteSampleCount() {
         this.completeSampleCount = completeSampleCount;
     }
+    
+    public String getReferenceGenome() { return  referenceGenome; }
```
agree, but I am not the one added setCompleteSampleCount(). I followed the rest is "one line" style.

khzhu commented on model/src/main/java/org/cbioportal/model/Mutation.java:
```diff
@@ -14,6 +14,7 @@
     private Integer normalAltCount;
     private Integer normalRefCount;
     private String aminoAcidChange;
+    private String chr;
```
**_chr_** was recorded in the mutation_event table, but never used. Instead, we have been calculated **_chr_** from the cytoband in the gene table, which will no longer work when we have multiple genomes. That is why this PR proposed to use **_chr_** recorded in the mutation_event table.

added chr JSON property to MutationMixin class.

khzhu commented on model/src/main/java/org/cbioportal/model/Mutation.java:
```diff
@@ -99,6 +100,10 @@ public String getAminoAcidChange() {
         this.aminoAcidChange = aminoAcidChange;
     }
+    
+    public String getChr() { return chr; }
```
no change, since most of properties in mutation are now one line style. 
I might need to set a rule to avoid mixed style in one file.

inodb replied:
sorry, i don't really understand this one. It seems like everywhere in this file it is block style. Only the newly introduced changes are on a single line?

khzhu replied:
@inodb , good catch!  changed to block style.

khzhu replied:
@inodb @sheridancbio , changed.

inodb replied:
doesn't seem to be changed?

khzhu replied:
@inodb , will take a look.

khzhu commented on persistence/persistence-mybatis/src/main/resources/org/cbioportal/persistence/mybatis/DiscreteCopyNumberMapper.xml:
```diff
@@ -184,13 +183,16 @@
         INNER JOIN genetic_profile ON sample_cna_event.GENETIC_PROFILE_ID = genetic_profile.GENETIC_PROFILE_ID
         INNER JOIN gene ON cna_event.ENTREZ_GENE_ID = gene.ENTREZ_GENE_ID
+        INNER JOIN cancer_study ON cancer_study.CANCER_STUDY_ID = genetic_profile.CANCER_STUDY_ID
+        INNER JOIN reference_genome_gene ON reference_genome_gene.ENTREZ_GENE_ID = cna_event.ENTREZ_GENE_ID
```
ENTREZ_GENE_ID, REFERENCE_GENOME_ID are both primary keys should be automatically indexed in MySQL.

khzhu commented on model/src/main/java/org/cbioportal/model/ReferenceGenomeGene.java:
```diff
@@ -0,0 +1,90 @@
+        this.length = length;
+    }
+
+    public Long getStart() { return this.start; }
```
since most classes in model are one line style, changed setter/getter in this class to one line style as well.

khzhu commented on persistence/persistence-mybatis/src/main/resources/org/cbioportal/persistence/mybatis/MolecularProfileMapper.xml:
```diff
@@ -32,6 +32,7 @@
         </include>
         FROM genetic_profile
         INNER JOIN cancer_study ON genetic_profile.CANCER_STUDY_ID = cancer_study.CANCER_STUDY_ID
+        INNER JOIN reference_genome ON cancer_study.REFERENCE_GENOME_ID = reference_genome.REFERENCE_GENOME_ID
```
you are right, that might something I forgot to remove. Originally we decided to support multiple genomes at the profile level, but found too complicated and changed to per study per genome build.

khzhu commented on persistence/persistence-mybatis/src/main/resources/org/cbioportal/persistence/mybatis/ReferenceGenomeGeneMapper.xml:
```diff
@@ -0,0 +1,109 @@
+        FROM reference_genome_gene
+        INNER JOIN gene on gene.entrez_gene_id = reference_genome_gene.entrez_gene_id
+        <if test="genomeName != null">
+            INNER JOIN reference_genome ON reference_genome.REFERENCE_GENOME_ID = reference_genome_gene.REFERENCE_GENOME_ID
```
same as MolecularProfileMapper it includes a StudyMapper in details projection.
```
        <if test="projection == 'DETAILED'">
            ,
            <include refid="org.cbioportal.persistence.mybatis.StudyMapper.select">
                <property name="prefix" value="${prefix}cancerStudy."/>
            </include>
        </if>
```

khzhu commented on persistence/persistence-mybatis/src/test/java/org/cbioportal/persistence/mybatis/DiscreteCopyNumberMyBatisRepositoryTest.java:
```diff
@@ -1,8 +1,6 @@
-import org.cbioportal.model.CopyNumberCountByGene;
-import org.cbioportal.model.DiscreteCopyNumberData;
-import org.cbioportal.model.Gene;
+import org.cbioportal.model.*;
```
changed.

khzhu commented on persistence/persistence-mybatis/src/test/java/org/cbioportal/persistence/mybatis/DiscreteCopyNumberMyBatisRepositoryTest.java:
```diff
@@ -62,7 +63,7 @@ public void getDiscreteCopyNumbersInMolecularProfileBySampleListIdDetailedProjec
             .getDiscreteCopyNumbersInMolecularProfileBySampleListId("study_tcga_pub_gistic", "study_tcga_pub_all", 
                 entrezGeneIds, alterations, "DETAILED");
-
+        
```
removed.

khzhu commented on persistence/persistence-mybatis/src/test/java/org/cbioportal/persistence/mybatis/ReferenceGenomeGeneMyBatisRepositoryTest.java:
```diff
@@ -0,0 +1,109 @@
+/*
```
dropped header.

khzhu commented on persistence/persistence-mybatis/src/test/java/org/cbioportal/persistence/mybatis/ReferenceGenomeGeneMyBatisRepositoryTest.java:
```diff
@@ -0,0 +1,109 @@
+        Assert.assertEquals((Integer) 207, gene.getEntrezGeneId());
+        Assert.assertEquals("AKT1", gene.getHugoGeneSymbol());
+    }
+    
```
double blank lines removed.

khzhu commented on persistence/persistence-mybatis/src/test/resources/testSql.sql:
```diff
@@ -1,8 +1,11 @@
+INSERT INTO `reference_genome` VALUES (2, 'human', 'hg38', 'GRCh38', NULL, 'http://hgdownload.cse.ucsc.edu/goldenPath/hg38/bigZips', '2013-12-01 00:00:00');
+
+INSERT INTO cancer_study (CANCER_STUDY_ID,CANCER_STUDY_IDENTIFIER,TYPE_OF_CANCER_ID,NAME,SHORT_NAME,DESCRIPTION,PUBLIC,PMID,CITATION,GROUPS,STATUS,IMPORT_DATE,REFERENCE_GENOME_ID) VALUES(1,'study_tcga_pub','brca','Breast Invasive Carcinoma (TCGA, Nature 2012)','BRCA (TCGA)','<a href=\"http://cancergenome.nih.gov/\">The Cancer Genome Atlas (TCGA)</a> Breast Invasive Carcinoma project. 825 cases.<b ...
+INSERT INTO cancer_study (CANCER_STUDY_ID,CANCER_STUDY_IDENTIFIER,TYPE_OF_CANCER_ID,NAME,SHORT_NAME,DESCRIPTION,PUBLIC,PMID,CITATION,GROUPS,STATUS,IMPORT_DATE,REFERENCE_GENOME_ID) VALUES(2,'acc_tcga','acc','Adrenocortical Carcinoma (TCGA, Provisional)','ACC (TCGA)','TCGA Adrenocortical Carcinoma; raw data at the <A HREF="https://tcga-data.nci.nih.gov/">NCI</A>.',1,'23000897','TCGA, Nature 2012',' ...
```
let's add in the future then.

khzhu commented:
HI @sheridancbio and @pieterlukasse , I've addressed all your comments and pushed changes back to the repo. Thanks again for taking time to review the changes!

khzhu commented on persistence/persistence-mybatis/src/main/resources/org/cbioportal/persistence/mybatis/PatientMapper.xml:
```diff
@@ -20,6 +20,7 @@
     <sql id="from">
         FROM patient
         INNER JOIN cancer_study ON patient.CANCER_STUDY_ID = cancer_study.CANCER_STUDY_ID
+        INNER JOIN reference_genome ON cancer_study.REFERENCE_GENOME_ID = reference_genome.REFERENCE_GENOME_ID
```
because it uses StudyMapper.select.

khzhu commented on persistence/persistence-mybatis/src/main/resources/org/cbioportal/persistence/mybatis/SampleListMapper.xml:
```diff
@@ -30,6 +30,7 @@
         </include>
         FROM sample_list
         INNER JOIN cancer_study ON sample_list.CANCER_STUDY_ID = cancer_study.CANCER_STUDY_ID
+        INNER JOIN reference_genome ON cancer_study.REFERENCE_GENOME_ID = reference_genome.REFERENCE_GENOME_ID
```
because it has .StudyMapper.select., StudyMapper queries the name of reference genome.
```
        <if test="projection == 'DETAILED'">
            ,
            <include refid="org.cbioportal.persistence.mybatis.StudyMapper.select">
                <property name="prefix" value="${prefix}cancerStudy."/>
            </include>
        </if>
```

pieterlukasse commented on db-scripts/src/main/resources/migration.sql:
```diff
@@ -396,12 +396,12 @@ CREATE TABLE `reference_genome` (
+VALUES (1, 'human', 'hg19', 'GRCh37', 2897310462, 'http://hgdownload.cse.ucsc.edu/goldenPath/hg19/bigZips', '2009-02-01');
+INSERT INTO `reference_genome` 
+VALUES (2, 'human', 'hg38', 'GRCh38', 3049315783, 'http://hgdownload.cse.ucsc.edu/goldenPath/hg38/bigZips', '2013-12-01');
+-- INSERT INTO `reference_genome` 
```
@khzhu will you add these back?

khzhu replied:
@pieterlukasse , added mm10 back.

pieterlukasse commented on core/src/main/java/org/mskcc/cbio/portal/scripts/ImportExtendedMutationData.java:
```diff
@@ -146,7 +146,18 @@ public void importData() throws IOException, DaoException {
+                try {
+                    String referenceGenome = cancerStudy.getReferenceGenome();
+                    genomeBuildName = DaoReferenceGenome.getReferenceGenomeByGenomeName(referenceGenome).getBuildName();
+                } catch (NullPointerException e) {
```
when does this happen? Please don't catch it but add the appropriate if/else structure.

khzhu replied:
@pieterlukasse, it happens when reference genome information is not included in the meta_study file, since "reference genome" is an optional property. So, try & catch is more appropriate approach in this case.

khzhu commented on core/src/main/java/org/mskcc/cbio/portal/scripts/ImportExtendedMutationData.java:
```diff
@@ -146,7 +146,18 @@ public void importData() throws IOException, DaoException {
                 String[] parts = line.split("\t", -1 ); // the -1 keeps trailing empty strings; see JavaDoc for String
                 MafRecord record = mafUtil.parseRecord(line);
+                CancerStudy cancerStudy = DaoCancerStudy.getCancerStudyByInternalId(geneticProfile.getCancerStudyId());
+                String genomeBuildName;
```
very good point, @pieterlukasse . moved outside the loop.

khzhu commented:
@pieterlukasse , I've address all your comments and pushed changes back to the repo. Please let me know if anything else I missed. Thanks for taking time review the PR!

pieterlukasse commented on db-scripts/src/main/resources/migration.sql:
```diff
@@ -396,17 +396,17 @@ CREATE TABLE `reference_genome` (
     `ENTREZ_GENE_ID` int(11) NOT NULL,
     `REFERENCE_GENOME_ID` int(4) NOT NULL,
-    `CHR` varchar(4) DEFAULT NULL,
+    `CHR` varchar(5) DEFAULT NULL,
```
wait, you can't make changes to previous migration scripts. The way you have to do this is to create a new migration part that adjusts the size from 4 to 5. The same holds for the values (2897310462,  3049315783, 2652783500) above.

khzhu replied:
oops...changed, @pieterlukasse.
```
UPDATE `reference_genome` SET `GENOME_SIZE` = 2897310462 WHERE `NAME`='hg19';
UPDATE `reference_genome` SET `GENOME_SIZE` = 3049315783 WHERE `NAME`='hg38';
UPDATE `reference_genome` SET `GENOME_SIZE` = 2652783500 WHERE `NAME`='mm10';
ALTER TABLE `reference_genome_gene` MODIFY COLUMN `CHR` varchar(5);
```

pieterlukasse commented on core/src/main/scripts/importer/validateData.py:
```diff
@@ -4606,9 +4621,11 @@ def main_validate(args):
-        portal_instance.load_genome_info(args.portal_properties)
-
+    # specify species and genomic information
+    portal_instance.species = args.species
```
there is no check that verifies if the given species is the same as the species already loaded in the portal. The code should check this. We can support multiple genomes in the same portal, but multiple species is an error situation. It is also fine if this check is done only in the Java layer. But there should be a check. It can be an easy check like (pseudo-code):
```
ref_genome_last_study = get last study's reference genome
if ref_genome_last_study.species != current_species
    abort
```

khzhu replied:
what about you have a brand new database or load a new study? I think we should check against portal.properties file if needed.

pieterlukasse replied:
@khzhu if the DB is empty, then the check can be skipped.

khzhu replied:
Hi @pieterlukasse , I do not think the validator can retrieve old version of the study. So, the validation of species should be added to CancerStudyReader util script the one does importing. The new method checkSpecies will do the check.
```
    private static Boolean checkSpecies(String studyId, String genomeName) {
        if (genomeName == null || genomeName == "") {
            return true;
        }
        try {
            CancerStudy oldCancerStudy = DaoCancerStudy.getCancerStudyByStableId(studyId);
            ReferenceGenome referenceGenome = DaoReferenceGenome.getReferenceGenomeByGenomeName(
                                oldCancerStudy.getReferenceGenome());
            return referenceGenome.getGenomeName().equalsIgnoreCase(genomeName);
        } catch (DaoException | NullPointerException e) {
            return true;
        }
    }
```

khzhu commented on core/src/main/java/org/mskcc/cbio/portal/scripts/ImportExtendedMutationData.java:
```diff
@@ -142,11 +142,23 @@ public void importData() throws IOException, DaoException {
+                genomeBuildName = DaoReferenceGenome.getReferenceGenomeByGenomeName(referenceGenome).getBuildName();
+            } catch (NullPointerException e) {
+                genomeBuildName = GlobalProperties.getReferenceGenomeName();
+            }
```
oops...changed. thanks, @pieterlukasse!

pieterlukasse commented:
Thanks @khzhu . Most of my comments are now resolved. Please let me know once you address the last ones. Main ones that are still open:
- [ ] validation to enforce only 1 species in the portal
- [ ] validation of CHR and ENTREZ_GENE_ID combinations before loading into `mutation_event` and `cosmic_mutation` tables

khzhu commented:
Hi @pieterlukasse , I will add one species check. For your item 2, do we really need?
can @jjgao , please comment on item 2 from @pieterlukasse . Thanks!

jjgao commented:
@khzhu I think it would be nice to have item 2 in the validator for validating MAFs. But I don't see that we have to do it in this PR. Maybe create a separate issue for that? @pieterlukasse

jjgao commented:
Thanks, @khzhu . Could you please log it in another issue?

khzhu commented:
@jjgao , I might look into this if it is a quick fix since I am already here. Thanks!

pieterlukasse commented:
@khzhu adding the 2nd validation in a follow-up PR is also fine with me. It is an important validation I think,  since it looks like CHR is used in some places of mutationMapper.js (if this file is still used).

khzhu commented:
@jjgao @pieterlukasse , I've created a sub task for chr and entrez_gene_id validation. We will verify the combination of chr and entrez_gene_id in PR 6254. thanks!

ao508 commented:
@khzhu I was going to make another pass but it looks like there are a lot of conflicts to resolve. Would you mind resolving and letting me know when this PR is updated? Hoping to get this merged in soon. Thanks!

khzhu commented:
hi @ao508 , will address two comments from @pieterlukasse I missed and rebase to the latest. Thanks!

khzhu commented:
hi @pieterlukasse , can you point two comments (file and lines) I missed. The links you sent last time did not take me to the right spots. Thanks!

inodb commented on service/src/test/java/org/cbioportal/service/util/AlterationEnrichmentUtilTest.java:
```diff
@@ -120,7 +121,7 @@ public void createAlterationEnrichments() throws Exception {
         Assert.assertEquals((Integer) 3, alterationEnrichment2.getEntrezGeneId());
         Assert.assertEquals("HUGO3", alterationEnrichment2.getHugoGeneSymbol());
-        Assert.assertEquals("CYTOBAND3", alterationEnrichment2.getCytoband());
+        Assert.assertEquals(null, alterationEnrichment2.getCytoband());
```
this seems incorrect?

khzhu replied:
Hi @inodb , that is right. for alterations we will have to fill out cytobands in the frontend, since alterations have no reference genome information associated with them. We will have to rely on the reference genome information of the study to backfill cytobands in the frontend.

khzhu commented:
Hi @ao508 @sheridancbio @pieterlukasse @inodb ,
rebased to the latest rc, addressed all review comments and passed CI :smile:. Please resolve your comments and approve the PR or let me know anything I missed. Thanks again for all your time and suggestions!

ao508 commented on core/src/main/scripts/importer/validateData.py:
```diff
@@ -288,26 +288,35 @@ def __init__(self, cancer_type_dict, hugo_entrez_map, alias_entrez_map, gene_set
+                        
+        #Set defaults for genome version and species
+        self.__species = 'human'
+        self.__ncbi_build = '37'
```
should `37` be changed to `GRCh37` to be consistent with how we refer to ncbi builds in the java code? [see here](https://github.com/cBioPortal/cbioportal/pull/5891/files#diff-110497640cf5b4d5d5b73305afbd03caR72)

ao508 commented on db-scripts/src/main/resources/migration.sql:
```diff
@@ -396,12 +396,21 @@ CREATE TABLE `reference_genome` (
     UNIQUE INDEX `BUILD_NAME_UNIQUE` (`BUILD_NAME` ASC)
 );
 
+<<<<<<< HEAD
```
unmerged file conflict

ao508 commented on db-scripts/src/main/resources/migration.sql:
```diff
@@ -396,12 +396,21 @@ CREATE TABLE `reference_genome` (
 VALUES (2, 'human', 'hg38', 'GRCh38', NULL, 'http://hgdownload.cse.ucsc.edu/goldenPath/hg38/bigZips', '2013-12-01');
 INSERT INTO `reference_genome`
 VALUES (3, 'mouse', 'mm10', 'GRCm38', NULL, 'http://hgdownload.cse.ucsc.edu//goldenPath/mm10/bigZips', '2012-01-01');
+>>>>>>> resolving code review comments/suggestions
```
another conflict

ao508 commented on docs/Using-the-dataset-validator.md:
```diff
@@ -38,8 +38,12 @@ optional arguments:
+  -ucsc UCSC_BUILD_NAME, --ucsc_build_name UCSC_BUILD_NAME
+                        UCSC reference genome assembly name (default: assumed hg19)
+  -ncbi NCBI_BUILD_NUMBER, --ncbi_build_number NCBI_BUILD_NUMBER
+                        NCBI reference genome build number (default: assumed 37 for UCSC reference genome build hg19)
```
use `GRCh*`

ao508 commented on docs/Using-the-dataset-validator.md:
```diff
@@ -474,8 +481,12 @@ optional arguments:
+                        species information (default: assumed human)
+  -ucsc UCSC_BUILD_NAME, --ucsc_build_name UCSC_BUILD_NAME
+                        UCSC reference genome assembly name (default: assumed hg19)
+  -ncbi NCBI_BUILD_NUMBER, --ncbi_build_number NCBI_BUILD_NUMBER
```
same comment here

ao508 commented on service/src/main/java/org/cbioportal/service/util/AlterationEnrichmentUtil.java:
```diff
@@ -62,7 +62,7 @@
             alterationEnrichment.setEntrezGeneId(gene.getEntrezGeneId());
             alterationEnrichment.setHugoGeneSymbol(gene.getHugoGeneSymbol());
-            alterationEnrichment.setCytoband(gene.getCytoband());
+            //alterationEnrichment.setCytoband(gene.getCytoband());
```
can we remove this if it's not going to be used or is there a plan to replace it with the reference_genome_gene.CYTOBAND? If so then can you add a TODO here?

khzhu commented:
@ao508 , replace those 37 with GRCh37 in validator documents. Also, removed the comment. For alterations, cytobands will be filled by the frontend react module.

khzhu commented:
@ao508 @inodb , addressed all your new comments, please have your final review. thanks!
